packages/diagram-common/src/model/elements/editSpecification.ts

```diff
@@ -0,0 +1,145 @@
+     */
+    export function isConsistent(entries: (EditSpecificationEntry | undefined)[][]): boolean {
+        const entriesWithGroup = entries.flatMap((group, groupIndex) =>
+            group.map((entry) => (entry ? { entry, group: groupIndex } : undefined))
+        );
+        const definedEntries = entriesWithGroup.filter((entry) => entry !== undefined);
+        if (definedEntries.length != entriesWithGroup.length) {
+            return false;
+        }
+        definedEntries.sort((a, b) => a.entry.range[0] - b.entry.range[0]);
+        for (let i = 1; i < entries.length; i++) {
+            const { entry: entry, group } = definedEntries[i];
+            const { entry: lastEntry, group: lastGroup } = definedEntries[i - 1];
+            const currentStart = entry.range[0];
+            const lastStart = lastEntry.range[0];
+            if (currentStart === lastStart) {
+                if (lastEntry.type != entry.type) {
+                    return false;
+                }
+                if (lastGroup != group) {
+                    return false;
+                }
+                if (
+                    entry.type === "replace" &&
+                    (entry.range[1] !== lastEntry.range[1] || !isTemplateEqual(entry.template, lastEntry.template))
+                ) {
+                    return false;
+                }
+            } else if (currentStart < lastEntry.range[1]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Checks if two entries are equal
+     *
+     * @param a the first entry
+     * @param b the second entry
+     * @returns true if the entries are equal
+     */
+    export function isEntryEqual(a: EditSpecificationEntry, b: EditSpecificationEntry): boolean {
+        if (a.type !== b.type || a.range[0] !== b.range[0] || a.range[1] !== b.range[1]) {
+            return false;
+        }
+        return isTemplateEqual(a.template, b.template);
+    }
+
+    /**
+     * Checks if two templates are equal
+     *
+     * @param a the first template
+     * @param b the second template
+     * @returns true if the template entries are equal
+     */
+    function isTemplateEqual(a: TemplateEntry[], b: TemplateEntry[]): boolean {
+        if (a.length !== b.length) {
+            return false;
+        }
+        for (let i = 0; i < a.length; i++) {
+            const first = a[i];
+            const second = b[i];
+            if (typeof first === "string" && typeof second === "string") {
+                if (first !== second) {
+                    return false;
+                }
+            } else if (typeof first !== "string" && typeof second !== "string") {
+                if (first.range[0] !== second.range[0] || first.range[1] !== second.range[1]) {
+                    return false;
+                }
+            } else {
+                return false;
+            }
```

Comment:
this would be incorrect, as this would then also be executed when the two cases above are chosen, but do not return
one could make it correct by adding an else case for the two inner if's with a `break`, but I don't think that would be an improvement

Reply:
Ah true